### ISSUE TITLE
when completing, we should display documentation for initializing parameters from the dartdoc from associated fields

### ISSUE THREAD

**Issue by sethladd** [flutter/flutter-intellij#385]
Would love to see what all the options are:

![screen shot 2016-10-27 at 1 11 44 pm](https://cloud.githubusercontent.com/assets/5479/19765115/fd3bc8d8-9c46-11e6-9b38-e11a59372e9c.png)

**Comment by devoncarew** [comment]
I don't believe they show up until you hit F1.

**Comment by sethladd** [comment]
Here's what I see when I hit F1:

![screen shot 2016-10-27 at 1 16 03 pm](https://cloud.githubusercontent.com/assets/5479/19765231/8dd777a2-9c47-11e6-9688-30ca0c2dda44.png)

Maybe analyzer/dartdoc doesn't know how to show docs per named arg?

I think Matan filed a similar issue to this.

**Comment by sethladd** [comment]
cc @matanlurey

**Comment by matanlurey** [comment]
Yep, I'd like to see something like:

```
ThemeDate {
  /// Defines the ... for ...
  Brightness brightness,
}
```

and once you type `brightness: ...`, you'd see the docs for the parameter.

**Comment by sethladd** [comment]
I volunteer the Flutter team to add the dartdocs if the IDE showed them.

cc @Hixie

**Comment by Hixie** [comment]
This may also require dartdoc changes so that the docs appear in the documentation.

Ideally arguments of the kind this.foo would automatically defer to the foo property otherwise we're in for a staggering amount of duplication...

**Comment by devoncarew** [comment]
I updated the title. I think this is a request for:

- letting named params be dardoc'd
- having dartdoc recognize as such
- borrowing docs automatically from associated members
- have tooltips and completions show these docs
- changing the style guide to support this type of usage (currently, it specifies docing params using in-line prose in the method dartdoc)

**Comment by mit-mit** [comment]
Actually, I think this is a bug. The parameters often (if not always) have docs, IntelliJ just don't show them (perhaps as they are on the member the param is setting?).

For example, try to set the `style:` parameters on a `Text` widget. IntelliJ shows no documentation, but it's [right there](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/widgets/text.dart#L75):

```
  Text(this.data, {
    Key key,
    this.style,
    this.textAlign,
    this.softWrap,
    this.overflow,
    this.textScaleFactor,
    this.maxLines,
  }) : super(key: key) {
    assert(data != null);
  }

  /// The text to display.
  final String data;

  /// If non-null, the style to use for this text.
  ///
  /// If the style's "inherit" property is true, the style will be merged with
  /// the closest enclosing [DefaultTextStyle]. Otherwise, the style will
  /// replace the closest enclosing [DefaultTextStyle].
  final TextStyle style;
```

**Comment by devoncarew** [comment]
So, when displaying the completions for the ctor param `foo`, if it's auto-assigned to the field `foo` (`this.foo`), we should display the docs for the member `foo`.

**Comment by devoncarew** [comment]
Konstantin, I'll follow up with some context.

**Comment by scheglov** [comment]
[CL for Analysis Server](https://codereview.chromium.org/2683233002)

**Comment by scheglov** [comment]
OK, this information is now available starting from https://github.com/dart-lang/sdk/commit/0a09ff53c3ef84d67b4f25eb3c4fa1d441f1e42d, so someone working on Dart plugin can use it.

**Comment by devoncarew** [comment]
Awesome, thanks!

**Comment by scheglov** [comment]
[This CL](https://codereview.chromium.org/2681233005) make it possible for Dart plugin to get documentation for field formal parameters in hover, and, in theory, in completion.

The `analysis.getHover` request works.
![image](https://cloud.githubusercontent.com/assets/384794/22809598/6313649c-eee8-11e6-9c6a-4c6806237b50.png)


But during completion we don't get documentation because by some reason IDEA does not use the specified element's location.

```Dart
class A {
  /**
   * aaaaaaaa
   * bbbbbbbb
   */
  final int fffff;

  A(int aaa, {this.fffff});
}

main(A a) {
  new A(1, )
}
```


![image](https://cloud.githubusercontent.com/assets/384794/22809502/e58c7658-eee7-11e6-966b-86c68e126493.png)
```JSON
{
  "event": "completion.results",
  "params": {
    "id": "40",
    "replacementOffset": 124,
    "replacementLength": 0,
    "results": [
      {
        "kind": "NAMED_ARGUMENT",
        "relevance": 1060,
        "completion": "fffff: ",
        "selectionOffset": 7,
        "selectionLength": 0,
        "isDeprecated": false,
        "isPotential": false,
        "docSummary": "aaaaaaaa\nbbbbbbbb",
        "docComplete": "aaaaaaaa\nbbbbbbbb",
        "element": {
          "kind": "PARAMETER",
          "name": "fffff",
          "location": {
            "file": "/Users/scheglov/dart/test/bin/test.dart",
            "offset": 89,
            "length": 5,
            "startLine": 8,
            "startColumn": 20
          },
          "flags": 0,
          "returnType": "int"
        },
        "parameterName": "fffff",
        "parameterType": "int"
      }
    ],
    "isLast": true
  }
}
```

```JSON
{
  "id": "1043",
  "method": "analysis.getHover",
  "params": {
    "file": "/Users/scheglov/dart/test/bin/test.dart",
    "offset": 115
  },
  "clientRequestTime": 1486687776493
}
```

Note that `115` is the offset of `new A(1, )`, not of the parameter `fffff` at `89`.

**Comment by scheglov** [comment]
The CL [was landed](https://github.com/dart-lang/sdk/commit/663b7f665e63eb9062028022d47dd24c56b789ff). Changes to IDEA Dart plugin still should be made.

**Comment by devoncarew** [comment]
@alexander-doroshko, Konstantin did some work to improve the docs we return for named params. It looks like he hit an issue w/ IntelliJ:

> But during completion we don't get documentation because by some reason IDEA does not use the specified element's location.

Does his analysis makes sense? If so, I can open a tracking issue at youtrack w/ the above details. Thanks!

**Comment by mit-mit** [comment]
@devoncarew to create the YouTrack entry

**Comment by devoncarew** [comment]
Created here: https://youtrack.jetbrains.com/issue/WEB-25478.

**Comment by alexander-doroshko** [comment]
Fixed

**Comment by mit-mit** [comment]
Was this change in the IJ or the Dart plugin? Just trying to figure out when and where it will be available.

**Comment by devoncarew** [comment]
The Dart plugin, so likely available in 2017.3.

**Comment by alexander-doroshko** [comment]
Yes, available in 2017.3 EAP + compatible Dart plugin

**Comment by mit-mit** [comment]
yay, confirmed working in 2017.3!

![screen shot 2018-01-04 at 15 17 41](https://user-images.githubusercontent.com/13644170/34567419-7c2dd962-f162-11e7-8752-02306f97b84b.png)
